phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/PhoenixRegionServerEndpoint.java

```diff
@@ -0,0 +1,119 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.phoenix.coprocessor;
+
+import com.google.protobuf.RpcCallback;
+import com.google.protobuf.RpcController;
+import com.google.protobuf.Service;
+
+import java.io.IOException;
+import java.util.Collections;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.CoprocessorEnvironment;
+import org.apache.hadoop.hbase.coprocessor.RegionServerCoprocessor;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.apache.phoenix.cache.ServerMetadataCache;
+import org.apache.phoenix.cache.ServerMetadataCacheImpl;
+import org.apache.phoenix.coprocessor.generated.RegionServerEndpointProtos;
+import org.apache.phoenix.coprocessorclient.metrics.MetricsMetadataCachingSource;
+import org.apache.phoenix.coprocessorclient.metrics.MetricsPhoenixCoprocessorSourceFactory;
+import org.apache.phoenix.protobuf.ProtobufUtil;
+import org.apache.phoenix.util.ClientUtil;
+import org.apache.phoenix.util.SchemaUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This is first implementation of RegionServer coprocessor introduced by Phoenix.
+ */
+public class PhoenixRegionServerEndpoint
+        extends RegionServerEndpointProtos.RegionServerEndpointService
+        implements RegionServerCoprocessor {
+    private static final Logger LOGGER = LoggerFactory.getLogger(PhoenixRegionServerEndpoint.class);
+    private MetricsMetadataCachingSource metricsSource;
+    protected Configuration conf;
+
+    @Override
+    public void start(CoprocessorEnvironment env) throws IOException {
+        this.conf = env.getConfiguration();
+        this.metricsSource = MetricsPhoenixCoprocessorSourceFactory
+                                .getInstance().getMetadataCachingSource();
+    }
+
+    @Override
+    public void validateLastDDLTimestamp(RpcController controller,
+            RegionServerEndpointProtos.ValidateLastDDLTimestampRequest request,
+            RpcCallback<RegionServerEndpointProtos.ValidateLastDDLTimestampResponse> done) {
+        metricsSource.incrementValidateTimestampRequestCount();
+        ServerMetadataCache cache = getServerMetadataCache();
+        for (RegionServerEndpointProtos.LastDDLTimestampRequest lastDDLTimestampRequest
+                : request.getLastDDLTimestampRequestsList()) {
+            byte[] tenantID = lastDDLTimestampRequest.getTenantId().toByteArray();
+            byte[] schemaName = lastDDLTimestampRequest.getSchemaName().toByteArray();
+            byte[] tableName = lastDDLTimestampRequest.getTableName().toByteArray();
+            long clientLastDDLTimestamp = lastDDLTimestampRequest.getLastDDLTimestamp();
+            String tenantIDStr = Bytes.toString(tenantID);
+            String fullTableName = SchemaUtil.getTableName(schemaName, tableName);
+            try {
+                LOGGER.debug("Verifying last ddl timestamp for tenantID: {}, tableName: {}",
```

Comment:
This log will be very noisy.

Reply:
@tkhurana I can remove this one. We have a similar [log line client side](https://github.com/apache/phoenix/pull/1883/files#diff-dd4686c6fc4cfc76ac9f187f3c9cf25e171516197e11d4cdd3a2d43eb26c7f2aR113) as well to log which region server the client is sending the request to. Should we remove that one too?

Reply:
Yes. We already have a metric for how many validation requests a RS got.

Reply:
done